File: Ruby/Object-oriented-programming/Lab/Model/Entity/Collections/Stack/linked_stack.rb
class LinkedStack < StackInterface

  def initialize
    @front = nil
    # @end = Link.new nil,nil,nil
    @size = 0
  end

  def size
    @size
  end

  def add(element)
    push element
  end

  def push(element)
    if @size == 0
      @front = Link.new element, nil, nil
      @end = @front
    else
      @end.next_node = Link.new element, @end, nil
      @end = @end.next_node
    end
    @size += 1
  end

  def pop
    rm = @end
    if @size == 0
      return nil
    elsif @size == 1
      @front = nil
    else
      @end = @end.prev_node
      @end.next_node = nil
    end
    @size -= 1
    return rm
  end

  def peek
    @end.element
  end


end

File: Csharp/Lab8/Lab8/Lab8/cng/Model/Collections/neww/stack/LinkedStack.cs
using System.Collections.Generic;

namespace Lab8.cng.Model.Collections.neww.stack
{
    public class LinkedStack<T> : AbstractCollection<T>, IStackInterface<T>
    {
        
        private Link<T> _front, _end;

        private int _size;

        public LinkedStack() {
            _front = null;
            _end = null;
            _size = 0;
        }

        public LinkedStack(T []ts){
            _front = null;
            _end = null;
            _size = 0;
            foreach (T t in ts){
                Add(t);
            }
        }

        public override int Size()
        {
            return _size;
        }

        public override void Add(T t)
        {
            Push(t);
        }

        public override IEnumerator<T> GetEnumerator()
        {
            LinkedStack<T> stack = this;
            while (!stack.IsEmpty())
            {
                yield return stack.Pop();
            }
        }

        public T Peek(){
            return _end.Element;
        }
        
        public void Push(T t) {
            if (_size == 0) {
                _front = new Link<T>(t, null, null);
                _end = _front;
            } else {
                _end.NextNode = new Link<T>(t, _end, null);
                _end = _end.NextNode;
            }
            _size++;
        }
        
        public T Pop() {
            Link<T> rm = _end;
            if (_size == 0) {
                return default(T);
            } else if (_size == 1) {
                _front = null;
            } else {
                _end = _end.PrevNode;
                _end.PrevNode = null;
            }
            _size--;
            return rm.Element;
        }
        
        public override int GetHashCode() {
            const int hashconst = 31;
            int result = 1;
            foreach (object o in this) {
                result += result * hashconst + o.GetHashCode();
            }
            return result;
        }
        
    }
}